### PR TITLE
PT-7355: Add mutation for company registration

### PR DESCRIPTION
**Pull request by einlied753**
## Description
Add mutation for company registration and notification
### Jira-link:

 https://virtocommerce.atlassian.net/browse/PT-7355
 https://virtocommerce.atlassian.net/browse/PT-7386
 https://virtocommerce.atlassian.net/browse/PT-7380
 https://virtocommerce.atlassian.net/browse/PT-7374
 https://virtocommerce.atlassian.net/browse/PT-7348

### CODE REVIEW

File: src/VirtoCommerce.ProfileExperienceApiModule.Data/Aggregates/RegisterCompanyAggregate.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+using Microsoft.AspNetCore.Identity;
+using VirtoCommerce.Platform.Core.Security;
+
+namespace VirtoCommerce.ProfileExperienceApiModule.Data.Aggregates
+{
+    public class RegisterCompanyAggregate
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ProfileExperienceApiModule.Data/Aggregates/RegisterCompanyAggregate.cs:6]:
```suggestion
    public class RegisterCompanyResult
```

File: src/VirtoCommerce.ProfileExperienceApiModule.Data/Aggregates/RegisterCompanyAggregate.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    {
+        public virtual CustomerModule.Core.Model.Organization Company { get; set; }
+        public virtual CustomerModule.Core.Model.Contact Owner { get; set; }
+        public virtual ApplicationUser Account { get; set; }
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ProfileExperienceApiModule.Data/Aggregates/RegisterCompanyAggregate.cs:10]: Do we really need this model here?

File: src/VirtoCommerce.ProfileExperienceApiModule.Data/Commands/RegisterCompanyCommand.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+using VirtoCommerce.ProfileExperienceApiModule.Data.Models.RegisterCompany;
+
+namespace VirtoCommerce.ProfileExperienceApiModule.Data.Commands
+{
+    public class RegisterCompanyCommand : ICommand<RegisterCompanyAggregate>
+    {
+        public string StoreId { get; set; }
+        public Company Company { get; set; }
+        public Owner Owner { get; set; }
+        public ApplicationUser Account { get; set; }
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ProfileExperienceApiModule.Data/Commands/RegisterCompanyCommand.cs:14]: Do we really need this model here?

File: src/VirtoCommerce.ProfileExperienceApiModule.Data/Commands/RegisterCompanyCommandHandler.cs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            _notificationSearchService = notificationSearchService;
+            _notificationSender = notificationSender;
+        }
+
+        public virtual async Task<RegisterCompanyAggregate> Handle(RegisterCompanyCommand request,
+            CancellationToken cancellationToken)
+        {
+            var aggregate = new RegisterCompanyAggregate();
+
+            try
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ProfileExperienceApiModule.Data/Commands/RegisterCompanyCommandHandler.cs:70]: Use `CancellationToken` to implement rollback flow, remove try..catch and unnecessary `Exceptions`. 
Pseudocode:
```
await RegisterCompanyAsync(cancellationToken);
if (cancellationToken.IsCancelled)
    await RollBackMembersCreationAsync();
```

File: src/VirtoCommerce.ProfileExperienceApiModule.Data/Commands/RegisterCompanyCommandHandler.cs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                account.Roles = new List<Role> { maintainerRole };
+                aggregate.AccountCreationResult = await CreateAccountAsync(account);
+
+                if (!aggregate.AccountCreationResult.Succeeded)
+                {
+                    await RollBackMembersCreationAsync(aggregate);
+                    return aggregate;
+                }
+
+                aggregate.Account = account;
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ProfileExperienceApiModule.Data/Commands/RegisterCompanyCommandHandler.cs:125]: Do we really need all data from the account?

File: src/VirtoCommerce.ProfileExperienceApiModule.Data/Commands/RegisterCompanyCommandHandler.cs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+            var ids = new[] { aggregate.Company?.Id, aggregate.Owner?.Id }
+                .Where(x => x != null)
+                .ToArray();
+            await _memberService.DeleteAsync(ids);
+
+            aggregate.Company = null;
+            aggregate.Owner = null;
+        }
+
+        private async Task<IdentityResult> CreateAccountAsync(ApplicationUser account)
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ProfileExperienceApiModule.Data/Commands/RegisterCompanyCommandHandler.cs:152]: Move out from to a service.

File: src/VirtoCommerce.ProfileExperienceApiModule.Data/Commands/RegisterCompanyCommandHandler.cs [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+            return result;
+        }
+
+        private async Task SetDynamicPropertiesAsync(IList<DynamicPropertyValue> dynamicProperties, IHasDynamicProperties entity)
+        {
+            if (dynamicProperties?.Any() ?? false)
+            {
+                await _dynamicPropertyUpdater.UpdateDynamicPropertyValues(entity, dynamicProperties);
+            }
+        }
```
Comment (Vectorfield4) [inline src/VirtoCommerce.ProfileExperienceApiModule.Data/Commands/RegisterCompanyCommandHandler.cs:182]: No reason for awaiting here.
```suggestion
        private Task SetDynamicPropertiesAsync(IList<DynamicPropertyValue> dynamicProperties, IHasDynamicProperties entity)
        {
            if (dynamicProperties?.Any() ?? false)
            {
                return _dynamicPropertyUpdater.UpdateDynamicPropertyValues(entity, dynamicProperties);
            }
            
            return Task.CompletedTask;
        }
```

**Review by Vectorfield4 — CHANGES_REQUESTED** [review]

**Review by Vectorfield4 — DISMISSED** [review]

**Review by akak1977 — APPROVED** [review]

**Pull request merged**